### DISCUSSION TITLE
Documentation for compilation

### DISCUSSION THREAD

**Discussion by spurin** [deskflow/deskflow#7504] [General]
Howdy all,

Great to see this discussion forum and the positivity in the community focus from @nbolton - As both a user of Synergy 3 Pro and someone who wishes to have hot fixes or cutting edge changes from Synergy 1 prior to them making it upstream, I strongly approve.

Nick, we chatted recently regarding issue: https://github.com/symless/synergy/issues/4245 and at some point, you've agreed to share a temporary binary.

I am however still frustrated that I wasn't able to compile the code base as desired on Windows and hence, this discussion forum is hopefully an opportunity to not only progress this but to also empower others so that they can easily contribute to the Synergy project.

The build guide at the moment, constructively, is somewhat presumptive for a new user and lacks detail, there's a variety of different routes that may result in different outcomes (my own experience). For context, am referring to this guide: https://github.com/symless/synergy/wiki/Build-Guide

Assuming that we have a brand new Windows 11 system, no software. The guide suggests running the following -

```
python scripts/install_deps.py
```

It's not clear, where or how this should be run. The slash pointing to the right (/) indicates that it is to be run in a unix like environment rather than via a command prompt (\\). There is also a requirement that python is available. Given this information, the route I initially started to follow was setting up WSL with ubuntu which, because of the detected system being linux, ended up seeming like the wrong path.

Second approach, command prompt. I installed Python using a release from https://www.python.org/downloads/windows/

After attempting to run the install_deps I had complaints that git wasn't installed. In a linux like environment I would typically apt install this but instead, opted for the windows binary from https://git-scm.com/

Tried again and seemingly got further. I did then however hit a failure point where the clang cl.exe executable was not in my path. This was quite confusing and eventually I found a bat file in a Visual Studio path which set this.

Finally, I ran the build command which seemed to succeed but alas, it didn't seem to run for enough time to compile anything and there wasn't anything in the output build folder.

How exactly do we build synergy from scratch, on a fresh windows systems?

What are the prerequisites, what assumptions are being made (for example, run this in WSL ubuntu, command prompt, powershell, vs code), what are the steps and what is the expected output.

If you're able to provide this and I'm able to get it working, I'll see about a pull request later on to improve the documentation.

Thanks

**Comment by nbolton** [comment]
> Great to see this discussion forum and the positivity in the community focus from @nbolton - As both a user of Synergy 3 Pro and someone who wishes to have hot fixes or cutting edge changes from Synergy 1 prior to them making it upstream, I strongly approve.

Really glad to hear that!

> The build guide at the moment, constructively, is somewhat presumptive for a new user and lacks detail, there's a variety of different routes that may result in different outcomes (my own experience). For context, am referring to this guide: https://github.com/symless/synergy/wiki/Build-Guide

I am inclined to agree. I think the problem here is that I have attempted to over-simplify what is a complicated problem and that just ends up wasting time for people who don't know about all of the pitfalls. The build instructions are a bit to reactive rather than comprehensive. My vision was that you'd just run the Python script to install the deps, and you'd be done. While this works on Unix-like OS, it's totally out of the question on Windows.

The solution is probably to flesh out the `BUILD.md` file a bit more for Windows users and then point the `README.md` and build guide at that file. I think that's step 1, and then step 2 will be to iterate based on your experience trying to follow the more comprehensive guide. However, at a high level: it should just require Python and Git, and you should be able to run it from any Windows shell (not WSL though) such as Command Prompt or PowerShell.

So that I can make sure it's included in the revised build instructions, what is the problem that you're now encountering?

**Reply by nbolton** [reply]
Or... perhaps I could create a `.cmd` and/or `.ps1` file that installs Python and Git for you, and then calls the Python script we have now. I decided not to write the whole thing as a `.ps1` file as we did that before and it made my eyes bleed.

**Reply by spurin** [reply]
I think having the base information and direction should be a good start, for example, install Python for Windows, install Git for Windows, use the command prompt, clone the project, run install_deps.py and then run the build command.

From your experience, the missing paths for cl.exe, is this something you've encountered or was this an unforseen issue.

I have a fresh build of windows on an ssd that I can clone and re-use so will give this another try and see how it gets on. If theres any other areas to be aware of then please let me know.

**Reply by nbolton** [reply]
> From your experience, the missing paths for cl.exe, is this something you've encountered or was this an unforseen issue.

Try restarting the shell that you're in (e.g. Command Prompt), delete the `build` directory, and re-run the configuration step.

**Comment by nbolton** [comment]
> cutting edge changes

Oh and in case you weren't aware, if we pushed a change recently from a PR (in the last 3 days), you can grab a build from the [CI workflow](https://github.com/symless/synergy/actions/workflows/ci.yml). These are of course bleeding edge changes and will likely contain bugs. Click any one of the successful ones and scroll down to 'Artifacts'. Windows is usually down there right at the bottom:
![image](https://github.com/user-attachments/assets/bc8e88a9-cae5-4da9-846c-409801384ea7)

**Comment by spurin** [comment]
First attempt, Fresh Windows 11 Installation -

1. Installed Python from main python website, required a reboot to work in command prompt (without, running python opened the windows store)
2. Installed git for Windows
3. From a command prompt, cloned the synergy source code using git
4. Changed to the synergy directory and run python scripts\install_deps.py

Downloaded and installed chocolatey, ran for a while, eventually failed 'Unable to find a valid Visual Studio instance'
Meanwhile in the background, chocolately failed with "the choco command was still not found, please re-run the this script"

Re-ran the script, chocolately complains that a pending system reboot request has been detected, however this is being ignored due to the current Chocolatey configuration.

This time, it appears to be installing VisualStudio2022 and DotNet3.5 - DotNet3.5 failed with a large output and the message that DotNet3.5 was NOT Successful, this processed other items further until eventually visual studio 2022 failed and indicated that a reboot is needed.

The main window continued, downloading various packages, until it indicated that the terminal/ide may need to be restarted to pick up the new environment variables.

Rebooted

Re-running script, states that visual studio and wixtoolset are already installed (although we know that Visual Studio and Dot Net Failed). The other dependencies appear to be satisfied.

Advice on this?

**Reply by nbolton** [reply]
Wow, that sounds like a horrific experience. Though, these do sound like typical Windows development woes, sadly. I'll have to try a fresh install on a VM myself.

Excellent recording of steps by the way.

Those VC++ errors are a pain. I have experienced having to uninstall and reinstall VC++ quite a lot on some occasions. For me, the install deps script always installs VC++ properly, but I did manage to get it in a weird state at some point so I had to manually remove everything Visual Studio related and re-install the [VS Community](https://visualstudio.microsoft.com/vs/community/) manually and select the VC++ development module.

**Reply by spurin** [reply]
I've installed VS Community, I selected DotNet and the VC++ Development module.

I found that I had to run the vcvarsall.bat file manually to get cl.exe in my path, in my case it was "c:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\Build\vcvarsall.bat"

With refactoring this process, I suggest keeping it as simple as possible, even if there are manual steps for those on Windows.

Manually install -

- Python (Python.org release for Windows)
- Git for Windows
- Visual Studio Community 2022, select DotNet and C++ Visual Studio Build Tools (Start Visual Studio then extit)

Then have the script do the rest.

Currently, I appear to have this running the build command `cmake -B build --preset=windows-release` albeit with some warnings, not sure if these are important or note. Summarised as -

- Check for working CXX compiler (cl.exe) - skipped
- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
- Looking for pthread_create in pthreads - not found

At this point, it says that Build file have been written to the build folder and I can see that the build folder is populated.

What I was ideally looking for was to build a synergy binary and the CI workflow appears to have additional steps, for example, the setup and use of ninja.

This got me thinking, it may be easier to have the Github CI workflow handle this. Do you use custom runners or do you make use of the ones with Github? I've implemented the change which we discussed in my own fork but at present, the runner is queued. Not sure if this is a Github issue or relates to the CI setup - https://github.com/spurin/synergy/actions/runs/10827402846

**Reply by nbolton** [reply]
> Do you use custom runners or do you make use of the ones with Github? I've implemented the change which we discussed in my own fork but at present, the runner is queued. Not sure if this is a Github issue or relates to the CI setup - https://github.com/spurin/synergy/actions/runs/10827402846

Good question. I made a new wiki page: [How to fork](https://github.com/symless/synergy/wiki/How-to-fork)

It has some detail about building packages on your own fork.

**Reply by nbolton** [reply]
> I found that I had to run the vcvarsall.bat file manually to get cl.exe in my path, in my case it was "c:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\Build\vcvarsall.bat"

Yeah, that can be tricky. Microsoft tries to make this easy by adding a shortcut in your start menu called the "[Developer Command Prompt for VS 2022](https://learn.microsoft.com/en-us/cpp/build/building-on-the-command-line?view=msvc-170)" but not everyone knows it's there. This would benefit from being documented in our instructions.

> * Python (Python.org release for Windows)
> * Git for Windows
> * Visual Studio Community 2022, select DotNet and C++ Visual Studio Build Tools (Start Visual Studio then extit)
> 
> Then have the script do the rest.

Sounds good, though the installing Visual Studio Community 2022 step shouldn't be strictly necessary, since the `visualstudio2022buildtools` Chocolatey package should take care of this. Perhaps we should simply say something along those lines: "You shouldn't need to install Visual Studio Community 2022 manually, but if you do..."

> Currently, I appear to have this running the build command `cmake -B build --preset=windows-release` albeit with some warnings, not sure if these are important or note. Summarised as -
> 
> * Check for working CXX compiler (cl.exe) - skipped

That's weird. I'm not sure why CMake is skipping that.

> * Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
> * Looking for pthread_create in pthreads - not found

Yeah those always bothered me on Windows but they seem innocuous. I think perhaps those come from some external dependency that we're relying on. I believe they originate from the Qt `find_package` call in `Libraries.cmake`.

> At this point, it says that Build file have been written to the build folder and I can see that the build folder is populated.

Congratulations! :tada: 

> What I was ideally looking for was to build a synergy binary and the CI workflow appears to have additional steps, for example, the setup and use of ninja.

The `install_deps.py` script (which is used for both setting up your dev env and setting up CI) should install Ninja with WinGet when not running in CI mode (since the GitHub runners don't and probably [wont support WinGet](https://github.com/actions/runner-images/issues/8584#issuecomment-1768217408)).

I hope I got all your points! Let me know if you have more questions :+1:

**Comment by spurin** [comment]
Super,Do you need anything from my current fork, if not, I’ll reset and try this out.

**Reply by nbolton** [reply]
I think perhaps I'm good from here. I merged https://github.com/symless/synergy/pull/7509 into `master`, then merged `master` in to your PR (https://github.com/symless/synergy/pull/7508) and resolved the merge conflicts. Next step is to add some args so those values can be tweaked. I'll give that some thought.

**Comment by sithlord48** [comment]
@nbolton We no longer have these build scripts and have new [build documents](https://github.com/deskflow/deskflow/wiki/Building) in the wiki perhaps its time to close this discussion ?

**Reply by nbolton** [reply]
Yeah, I believe this is satisfied now that we got rid of the scripts and replaced them with [docs](https://github.com/deskflow/deskflow/wiki/Building).
